### ISSUE TITLE
Null reference error on 2nd reload when IL editing with references

### ISSUE THREAD

**Issue by Spiky-73** [bluemagic123/tModLoader#3461]
### Version

Latest 1.4-stable

### OS

Windows

### Platform

Steam

### This bug affects

Mod capability as a Modder

### Description

The games throws a `System.NullReferenceException` when reloading a mod for the 2nd time if this mods IL edits a function and emits a delegate using a `ref` (e.g. `cursor.EmitDelegate((bool res, ref DropAttemptInfo info) => { ... })`.

### Log File

[client.log](https://github.com/tModLoader/tModLoader/files/11415992/client.log)

### Steps to reproduce

A mod containing tihs file triggers the error
```csharp
using System.Reflection;
using Mono.Cecil.Cil;
using MonoMod.Cil;
using Terraria.GameContent.ItemDropRules;
using Terraria.ModLoader;

namespace Testing;
public class Testing : Mod {

    public static Testing Instance { get; private set; } = null!;

    public override void Load() {
        Instance = this;
        IL.Terraria.GameContent.ItemDropRules.ItemDropResolver.ResolveRule += ILResolveRule;
    }

    public override void Unload() {
        Instance = null!;
    }

    private static void ILResolveRule(ILContext il) {
        const byte ArgDropAttemptInfo = 2;

        ILCursor cursor = new(il);
        MethodInfo candropMethod = typeof(IItemDropRule).GetMethod(nameof(IItemDropRule.CanDrop), BindingFlags.Public | BindingFlags.Instance)!;
        if (!cursor.TryGotoNext(i => i.MatchCallvirt(candropMethod))) {
            Testing.Instance.Logger.Error($"\"{nameof(ItemDropResolver)}.ResolveRule\" il hook could not be applied");
            return;
        }
        cursor.GotoPrev().GotoPrev();
        cursor.Emit(OpCodes.Ldarga_S, ArgDropAttemptInfo);
        cursor.EmitDelegate((ref DropAttemptInfo info) => {
            Instance.Logger.Debug("delegate called");
        });
    }
}
```

1. Launch TML normaly
2. (Enter a world and kill an ennemy)
3. Force reload the mods
4. Enter a world and kill an ennemy

### Expected Behavior

`delegate called` should be logged in the console when an ennemy is killed.

### Actual Behavior

`delegate called` is logged after the initial load.
After a Force Reload, the games throws a `System.NullReferenceException` when an ennemy is killed and ennemies stop dropping items.

### Reproduction frequency

Always with the code above.

### Additional Information

Unloading the hook manualy does not help.

I have no clue as for why it happens only with refs. delegates without them work fine, even after a reload.

A workaround is to use a delegate taking a value and returning one rather than taking a ref.

**Comment by JavidPack** [comment]
1.4.4, the next version of tmod, underwent large monomod changes, so you'll want to test there. Once you do, let us know.

This likely won't be able to be fixed on 1.4, since monomod has changed so much since then, a fix in monomod won't be easy to backport to the version of monomod 1.4 uses.

**Comment by JavidPack** [comment]
Closing, inactive.
